specifications/secure-launch-specification.rst

```diff
@@ -0,0 +1,719 @@
+        u16 flags;
+        u16 reserved;
+        u64 entity;
+        u64 size;
+        char evt_info[TPM_EVENT_INFO_LENGTH];
+    };
+
+DRTM Policy Entry Entity Types
+''''''''''''''''''''''''''''''
+
+The list of valid entity types for DRTM Policy entries.
+
+.. code-block:: c
+    :linenos: 1
+
+    #define SLR_ET_UNSPECIFIED  0x0000
+    #define SLR_ET_SLRT         0x0001
+    #define SLR_ET_BOOT_PARAMS  0x0002
+    #define SLR_ET_SETUP_DATA   0x0003
+    #define SLR_ET_CMDLINE      0x0004
+    #define SLR_ET_EFI_MEMMAP   0x0005
+    #define SLR_ET_RAMDISK      0x0006
+    #define SLR_ET_TXT_OS2MLE   0x0010
+    #define SLR_ET_UNUSED       0xffff
+
+DRTM Policy Entry Flags
+'''''''''''''''''''''''
+
+The list of valid flags for DRTM Policy entries.
+
+.. code-block:: c
+    :linenos: 1
+
+    #define SLR_POLICY_FLAG_MEASURED	0x1
+    #define SLR_POLICY_IMPLICIT_SIZE	0x2
+
+Intel TXT Platforms
+~~~~~~~~~~~~~~~~~~~
+
+When on Intel platforms specific information needs to be conveyed to Secure
+Launch.
+
+Intel TXT Info
+^^^^^^^^^^^^^^
+
+:REQUIRED: This entry **MUST** be present on Intel platforms.
+
+Intel TXT requires for the pre-launch environment to pass MSR and MTRR state
+across to the post-launch environment.
+
+:tag: SLR_ENTRY_INTEL_INFO
+:saved_misc_enable_msr: Saved MSR values
+:saved_bsp_mtrrs: Saved BSP MTRRs
+
+.. code-block:: c
+    :linenos: 1
+
+    struct slr_entry_intel_info {
+        struct slr_entry_hdr hdr;
+        u64 saved_misc_enable_msr;
+        struct txt_mtrr_state saved_bsp_mtrrs;
+    };
+
+Saved MTRR State
+""""""""""""""""
+
+:code:`struct txt_mtrr_state`
+
+:default_mem_type: The default memory type for regions not covered by an MTRR
+:mtrr_vcnt: Number of variable MTRR pairs in the mtrr_vcnt array
+:mtrr_pair: Array of variable MTRR pairs to restore post launch
+
+:code:`struct txt_mtrr_pair`
+
+:mtrr_physbase: Physical base address for variable MTRR
+:mtrr_physmask: Physical mask for the variable MTRR
+
+.. code-block:: c
+    :linenos: 1
+
+    struct txt_mtrr_pair {
+        u64 mtrr_physbase;
+        u64 mtrr_physmask;
+    };
+
+    struct txt_mtrr_state {
+        u64 default_mem_type;
+        u64 mtrr_vcnt;
+        struct txt_mtrr_pair mtrr_pair[TXT_VARIABLE_MTRRS_LENGTH];
+    };
+
+AMD Secure Launch Platforms
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+AMD SKINIT Info
+^^^^^^^^^^^^^^^
+
+A placeholder for info specific to AMD SKINIT.
+
+.. code-block:: c
+    :linenos: 1
+
+    struct slr_entry_amd_info {
+        struct slr_entry_hdr hdr;
+    };
+
+ARM DRTM Environments
+~~~~~~~~~~~~~~~~~~~~~
+
+ARM Info
+^^^^^^^^
+
+A placeholder for info specific to ARM DRTM environments.
+
+.. code-block:: c
+    :linenos: 1
+
+    struct slr_entry_arm_info {
+        struct slr_entry_hdr hdr;
+    };
+
+UEFI Environments
+~~~~~~~~~~~~~~~~~
+
+To support UEFI bootloaders that may do additional configuration of the Secure
+Launch kernel, the EFI SLRT entries provide a means to convey any operational
```

Comment:
The name came from the fact that Linux calls it the efi-stub, and not the uefi-stub. That said, I have no qualms changing it to UEFI Info/Config/Config Entry.

Reply:
In some places, we try to be scientifically precise; in others, we pass. What is more, we mix UEFI and EFI, and we should explain that either be consistent. The key question is for whom we write this spec because we should understand the reader.

I had this discussion with Xeno, and he insisted that in Apple security documentation, the UEFI term will be used despite all other Apple documentation using EFI. The fact that one uses incorrect terminology should not be the reason for using an incorrect one by us.

I will disagree and commit if you insist, but at least my point would be recorded.

Reply:
No need, I made the change already. As for Apple, there is a very good reason for Xeno to be explicit because prior to UEFI, Apple did in fact use the Open Firmware EFI. On x86, only a few niche environments had EFI, Sun x86 boxes and Itaniums (aka Itanics) are the ones I can recall.